T5440: Restore pre/postconfig scripts if user deleted them

Apachez- opened the pull request:
<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Change Summary
<!--- Provide a general summary of your changes in the Title above -->
If user deletes pre/postconfig scripts from directory /config/scripts then its not obvious how to restore the empty editions since they will only show up on a fresh install.

Bugfix is to adjust /usr/libexec/vyos/init/vyos-router so except for copying the empty scripts (if scripts are missing in /config/scripts) it will also set proper group and permission.

This will also fix the issue where preconfig script doesnt show up after an upgrade (add system image) as described in: https://vyos.dev/T5436

## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
<!-- All submitted PRs must be linked to a Task on Phabricator. -->
* https://vyos.dev/T5440

Also fixes:
* https://vyos.dev/T5436

## Component(s) name
<!-- A rather incomplete list of components: ethernet, wireguard, bgp, mpls, ldp, l2tp, dhcp ... -->
boot

## Proposed changes
<!--- Describe your changes in detail -->
1) Edited /usr/libexec/vyos/init/vyos-router.

2) Added functions "restore_if_missing_preconfig_script" and "restore_if_missing_postconfig_script".

3) In start-function made sure that "restore_if_missing_preconfig_script" runs ahead of "run_preconfig_script" as with "restore_if_missing_postconfig_script" runs ahead of "run_postconfig_scripts".

## How to test
<!---
Please describe in detail how you tested your changes. Include details of your testing
environment, and the tests you ran. When pasting configs, logs, shell output, backtraces,
and other large chunks of text, surround this text with triple backtics
```
like this
```
-->
1) Delete the files "/config/scripts/vyos-preconfig-bootup.script" and "/config/scripts/vyos-postconfig-bootup.script" and reboot.

2) Verify that the original (empty) files have been recreated in /config/scripts with proper group and permissions after reboot.

3) Edit both script-files and add "touch /config/scripts/test_preconfig.txt" to the preconfig-script and "touch /config/scripts/test_postconfig.txt" to the postconfig-script and reboot.

4) Verify that files "/config/scripts/test_preconfig.txt" and "/config/scripts/test_postconfig.txt" have been created after reboot but also that the content of pre/postconfig-scripts havent been overwritten.

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [ ] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

Apachez- committed:
T5440: Restore pre/postconfig scripts if user deleted them

sarthurdev commented on src/init/vyos-router:
```diff
@@ -101,6 +101,16 @@ load_bootfile ()
+restore_if_missing_preconfig_script ()
+{
+    if [ ! -x $vyatta_sysconfdir/config/scripts/vyos-preconfig-bootup.script ]; then
+        cp /usr/lib/live/mount/rootfs/*/opt/vyatta/etc/config/scripts/vyos-preconfig-bootup.script $vyatta_sysconfdir/config/scripts/
```
Why copy each file across all installed images to the current one?

Apachez- replied:
Because according to the documentation these files (preconfig- and postconfig-script) are part of VyOS:

https://docs.vyos.io/en/latest/automation/command-scripting.html

By default each file is empty (except for the comment of the purpose of each file).

If user removes them then they should be recreated (empty) upon next reboot.

If the files already exists at destination then they wont be overwritten.

This fix also fixes the fact that preconfig-script never shows up after a system upgrade (add system image) since the preconfig-code was previously broken (it was called for in the init-script but the file never existed which this fixes that case aswell).

The use of wildcard is because the empty pre-/postconfig-script should exist in all installed images (this is the rootfs aka the squashfs file from the iso which is being referenced) so which one is selected doesnt matter primarly.

But sure if there is some way to select the last available then this would be prefered.

Also there is missing a environment variable that points to current rootfs so if that can be setup then of course that env variable can be used instead.

sarthurdev replied:
> If user removes them then they should be recreated (empty) upon next reboot.

I disagree, they aren't essential - if they're missing, nothing executes. They can be added manually if needed.

> The use of wildcard is because the empty pre-/postconfig-script should exist in all installed images (this is the rootfs aka the squashfs file from the iso which is being referenced) so which one is selected doesnt matter primarly.

However an old script from a previous image could then be copied to the newly installed image and execute commands that are no longer relevant/could harm the current configuration.

If you wanted to guarantee the file is created empty, you could instead use `touch <path>` to generate the file.

c-po replied:
I aggree - but we should not only `touch(1)` the file if it‘s missing, but also create it with an appropriate note at the top.

Something like:

```bash
FOOFILE=$vyatta_sysconfdir/config/scripts/vyos-preconfig-bootup.script
if [ ! -x $FOOFILE ]; then
    cat << EOF > $FOOFILE
#!/bin/sh
# VyOS preconfig script
# add content here foo - bar baz
    EOF
    chmod +x $FOOFILE
fi
```

Potentially we could remove the entire template from the ISO script, too and just rely on this line to create the script in any case.

Apachez- replied:
@sarthurdev They are essential since they are mentioned in the docs regarding on how to perform post- and preconfig scripts. So if the native files are missing they should be recreated from the original "empty" ones.

Since the files are created by /src/debian/vyos-1x.postinst and are part of the squashfs they should be copied instead of having two places where the content of the files exists. This way the docs also match the reality.

@c-po I dont think the content of the files should be different, it should match the headers as displayed in the docs (https://docs.vyos.io/en/latest/automation/command-scripting.html). If something then a logging entry (also pushed to console) could be performed.

Apachez- replied:
I have created following feature request regarding current rootfs dir path that would resolve the use of wildcard as with this pull request: https://vyos.dev/T5457

Apachez- replied:
PR created for T5457: https://github.com/vyos/vyatta-op/pull/65

Apachez- replied:
I have added an update to this PR2135: https://github.com/vyos/vyos-1x/pull/2135/commits/d4d8517e0b78c8eae5d0b5518b09a0c4cb594e0e

Changes to original PR:

1) Using ${vyos_rootfs_dir} (now available thanks to merge of https://vyos.dev/T5457) instead of wildcard for if_missing_restore_preconfig and if_missing_restore_postconfig.

2) Adjusted chmod from 755 to 750 for the destination files (pre-/postconfig files copied to /config/scripts it missing).

Apachez- replied:
Hmm previous PR update needed a 2nd update but now I cant locate it :-S

Apachez- replied:
Found it!

This PR update2 contains following fixes compared to original PR2135:

https://github.com/vyos/vyos-1x/pull/2135/commits/2bfdfdfa80f9c351eaf57bffdc26ee772df8a59c

1) Using ${vyos_rootfs_dir} instead of wildcard for cp-operation in restore_if_missing_preconfig_script and restore_if-missing_postconfig_script.

2) Changed chmod from 755 to 750 for the files copied to destination.

3) Added proper brackets ${} for variables used by these two functions.

sarthurdev reviewed (commented):
I don't like the idea of recreating these files if they were purposefully deleted.

If the purpose is to fix a bug on upgrade, that issue should be directly addressed instead.

Apachez- committed:
Update vyos-router

Using variable ${vyos_rootfs_dir} instead of wildcard for both restore_if_missing_preconfig and restore_if_missing_postconfig.

Also adjusted chmod from 755 to 750.

Apachez- committed:
Update vyos-router

Fixed variables and the path for sourcefile of cp-operation.

sever-sever commented:
I have 2 scripts
```
vyos@r14:~$ sudo ls -la /config/scripts/vyos-postconfig-boot*
-rwxrwxr-x 1 root vyattacfg 413 Sep  6 15:04 /config/scripts/vyos-postconfig-boot.script
-rwxrwxr-x 1 root vyattacfg 230 Jun 27 06:17 /config/scripts/vyos-postconfig-bootup.script
vyos@r14:~$ 
```
Is it expected? Initially we use `vyos-postconfig-boot.script`

Apachez- commented:
According to https://github.com/vyos/vyos-1x/blob/current/src/init/vyos-router it should be named:

```
/config/scripts/vyos-preconfig-bootup.script
/config/scripts/vyos-postconfig-bootup.script
```

Which also matches the content of the manual:

https://docs.vyos.io/en/latest/automation/command-scripting.html?highlight=preconfig#preconfig-on-boot

Whats the content of that  /config/scripts/vyos-postconfig-boot.script ?

Could it be some leftover from previous install that is the timetag of "Sep 6 15:04" is when you did a add system image?

Anyhow unless you call that *boot.script from elsewhere it doesnt seem to be called by VyOS itself (which according to above source link is calling *bootup.script) so if you have a backup of that file it should be safe to delete it (again depending on its content, move the content to the *bootup.script file).
